keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,480 @@
+
+## Summary
+
+Currently, built-in API types are written as Go structures annotated
+with markers to add additional semantics. This method has inspired
+controller-tools who has re-used and improved on that technique for
+building CRDs. We're proposing using the `// +default` marker to
+declaratively define defaults (and have kubebuilder use that as well)
+since that mechanism still doesn't exist for built-in. This proposal
+tries to solve that gap.
+
+## Motivation
+
+Having a new `// +default` marker for built-in types has several
+important benefits:
+
+- It's more convenient for types authors.
+- The types and their behaviors are easier to understand since they are
+described together.
+- They are less error-prone, and prevent people from using anti-patterns
+like "cross-field dependency defaults" (default depends on another
+field), non-deterministic or complicated defaults. These patterns
+might be useful in limited circumstances, but it should be easier to do
+the right thing in most cases.
+- Removes the inconsistency with CRDs that already support this
+mechanism.
+- Solves bug where associative keys need to either be required or
+  defaulted.
+
+### Goals
+
+The goal is to add a new `// +default` tag to our current built-in Go
+IDL. That tag will be transformed into the OpenAPI `default` tag and
+then routed to defaulting functions so that defaulting can be done
+declaratively.
+
+### Non-Goals
+
+While this proposal is probably paving the way for additional OpenAPI
+based validation, validation is entirely out-of-scope for this project.
+
+This proposal also doesn't intend to remove the existing defaulting
+mechanism since some defaulting still requires non-static values that
+need to be computed and can't be defined declaratively. We could replace
+some defaulting that doesn't require to be done as code though.
+
+## Proposal
+
+Since the goal of this proposal is to increase consistency with the CRD
+defaulting, we propose to use the same mechanism (both in OpenAPI and
+similar to what kubebuilder does). This defaulting will happen BEFORE
+the existing defaulting functions are called (existing functions can be
+updated to panic so that we can see if defaults are properly applied).
+
+One very important nuance with CRDs is that CRDs are deserialized in
+unstructured objects that can distinguish between specified and
+unspecified fields. That is not the case for built-in types that are
+directly deserialized into Go-struct, where the unspecified fields are
+zero-valued. It's important to understand that only zero-value
+omit-empty fields will be defaulted using that mechanism.
+
+Computed defaults will not be supported as a declarative marker and will
+continue being supported through the existing defaulting functions.
+
+### Risks and Mitigations
+
+The main risk is that we could end-up with different defaults for
+existing types which would be extremely confusing for our users.
+
+## Design Details
```

Comment:
**I am not convinced about the complexity-benefit ratio of this approach.** I like the idea to support `// +default`. But it would be much much easier probably to just generate defaulter funcs and use the mechanisms we have unchanged. Wiring server side apply into our serialization stack is huge. Getting defaulting semantics right (i.e. without difference to what we have for native types and for CRDs today) is hard.

@liggitt @lavalamp am I alone with this gut feeling, especially about the complexity-benefit ratio of wiring this up in our (de)serialization stack?

Reply:
I think I've addressed that comment. I agree with you and just had forgotten to update some sections of the document.

I'm planning on generating defaulting functions and keep the current mechanisms unchanged, this is just an additional layer.

Reply:
:+1: